### PR TITLE
Scripted 2D optics component

### LINKED ISSUE #1102: Feature request: 2D scopes framework
ACE3 and BWMod has sweet custom 2D scopes with bunch of features like recoil animations, reticle switching and so on.

Would be great to have similar framework in CBA and plug custom scopes with minimal set ups from my mod's side.

Expected Minimum: To have all needed settings, scripts, handlers and Rsc base classes in CBA + some documentation

### PR DESCRIPTION
**Pull request by commy2**
**When merged this pull request will:**
- ports the BWA3 scripted 2d optics "framework" (turns it into one) to CBA
- animated 2d scopes (recoil, zooming etc.)
- iluminated reticles during the night
- shows current magnification as ui element in the weapon info box
- toggle setting between PIP background and darkened background
- close #1102

- [x] docs
- [x] example configs
- [x] function headers, variable tags
- [x] the settings
- [x] support for multiple weapons with carry handle optics
- [x] move some compatiblity stuff to ACE (arsenal, create events in events)
- [x] only init pip camera after setting enables system once

[# Documentation #](https://github.com/CBATeam/CBA_A3/wiki/Scripted-2D-optics)

### CODE REVIEW

**Comment by commy2** [comment]
Reminder to squash.

File: addons/2doptics/fnc_currentOptic.sqf [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Internal Function: cba_2doptics_fnc_currentOptic
+
+Description:
+    Creates a new setting for that session.
```
Comment (dedmen) [inline addons/2doptics/fnc_currentOptic.sqf]: @todo?

File: addons/2doptics/reticles/pip.rvmat [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+VertexShaderID = "Basic";
+
+class Stage1 {
+    texture = "#(argb,8,8,3)color(0.5,0.5,0.5,0.5,dt)";
+    uvSource = "tex";
+
+    class uvTransform {
+        aside[] = {1,0,0};
+        up[] = {0,1,0};
+        dir[] = {0,0,0};
```
Comment (dedmen) [inline addons/2doptics/reticles/pip.rvmat]:
```suggestion
        dir[] = {0,0,1};
```

Comment (commy2) [reply addons/2doptics/reticles/pip.rvmat]: Explain?

Comment (dedmen) [reply addons/2doptics/reticles/pip.rvmat]: It's a matrix.
3 direction vectors (side,up,dir) and position.
0,0,0 is a invalid direction vector.

Comment (commy2) [reply addons/2doptics/reticles/pip.rvmat]: Makes sense, but why is it ZYX?

Comment (dedmen) [reply addons/2doptics/reticles/pip.rvmat]: You are asking me why arma does weird things?

Comment (commy2) [reply addons/2doptics/reticles/pip.rvmat]: XZY sounds more "normal" (for BI standards).

**Comment by commy2** [comment]
Release of this would require https://github.com/acemod/ACE3/pull/6921, or it will go wrong in the ACE Arsenal.

**Comment by Blutze** [comment]
As mentioned on the BWMod gitlab, unscoping currently resets magnification to the minimum. With the current state of the CBA weapon handling framework, this has to be triggered on every shot. Is there a way to prevent the magnification resetting, potentially extending to the vanilla Rangefinder and Laser Designators?

**Comment by commy2** [comment]
That is just what happens in the base game when you unscope, right? I don't think that can be changed, as there is no command to set the current magnification by script.

**Comment by 10Dozen** [comment]
> That is just what happens in the base game when you unscope, right? I don't think that can be changed, as there is no command to set the current magnification by script.

Engine based optics mode is saved for `modelOptics` and `discreteFov` when changed via Num+.

Maybe GUI optics controls may be binded to same key as optics zoom (default: Num+/Num-) and it will control GUI and engine part same time? It will require optics attachement handler to reset scripted state when optics detached and replaced with new one.

**Comment by commy2** [comment]
I don't understand how that is supposed to work. I have my doubts this can be achieved, but feel free to fork this branch and open a pull request...

**Comment by Blutze** [comment]
> That is just what happens in the base game when you unscope, right?

Nope, only on rangefinder and laser designators. The LRPS/Nightforce NXS for instance stays on full "25x" instead of resetting to "5x". Even when you have multiple different levels (the RKSL S&B5-25 has four magnification stages), they don't get reset.

**Comment by commy2** [comment]
That is not a zooming optic. It has different magnifications you can switch between. That is something completely different. Yes, the rangefinder is an example of how continuous zoom behaves in Arma and there is no way around it.

**Comment by commy2** [comment]
Zooming optics are the Rangefinder or every CQB sight.

**Comment by commy2** [comment]
Example:
The G36A2 is a weapon with carry handle optic and picatinny rail. We want to be able to attach the RSAS CQB sight on the rail and still be able to use the carry handle optic. We cannot however add the carry handle optic to the G36A2 and the RSAS, because then either the G36A2 would lose access to the carry handle optic when equipped with the RSAS, or other weapons equipped with the RSAS would gain access to the carry handle optic.
The solution is to switch around the RSAS versions by this framework to either be the normal CQB sight or the CQB sight with additional hidden carry handle optic mode to replace the carry handle optic mode of the G36A2 seamlessly to the user.
We also want a variant of the carry handle optic with PIP camera to emulate the scopes magnification while everything outside the scope remains unmagnified (depicted by a circular PIP field around the scope body).
Thus we need two weapons: the normal G36A2 (`BWA3_G36A2`) and a PIP version of the G36A2 (`BWA3_G36A2_pip`).
And we also need 3 optics: the normal RSAS (`BWA3_optic_RSAS`), the RSAS with the hidden G36A2 carry handle optic mode (`BWA3_optic_RSAS_G36A2`), and a RSAS with the PIP version of the carry handle optic mode (`BWA3_optic_RSAS_G36A2_pip`).

<details><summary>Weapon and Optic config</summary>

```cpp
// A base class for the optic slot config. All three optics have to be set compatible with the weapon or they will not be able to be attached by the game.
class BWA3_CowsSlot_CarryHandle {
    // These are nothing special.
    iconPicture = "\A3\Weapons_F\Data\UI\attachment_top.paa";
    iconPinpoint = "Bottom";
    iconPosition[] = {0.6,0.33};
    iconScale = 0.15;
    linkProxy = "\A3\data_f\proxies\weapon_slots\TOP";
    displayName = "$STR_A3_CowsSlot0";

    // The three optic versions. If they are not set to be compatible, the optic might disappear when they were supposed to be switched by the framework!
    class compatibleItems {
        BWA3_optic_RSAS = 1;
        BWA3_optic_RSAS_G36A1 = 1;
        BWA3_optic_RSAS_G36A1_pip = 1;
    };
};

// The base class for the carry handle optic. Make sure to set `useModelOptics` to 1. This was made a base class so both the weapon and the optic with hidden weapon specific optic mode can inherit it.
class BWA3_G36CarryHandleScope_base {
    opticsID = 1;
    useModelOptics = 1;
    opticsPPEffects[] = {"OpticsCHAbera5","OpticsBlur5"};
    opticsFlare = 1;
    opticsDisablePeripherialVision = 1;
    opticsZoomMin = "0.25/3";
    opticsZoomMax = "0.25/3";
    opticsZoomInit = "0.25/3";
    memoryPointCamera = "opticView";
    distanceZoomMin = 200;
    distanceZoomMax = 200;
    cameraDir = "";
    visionMode[] = {"Normal"};
};

// This contains all the config entries for the framework. Again a base class so the entries can be shared by weapon and special optic.
class BWA3_G36CarryHandleScriptedOptic_base {
    reticleTexture = "\bwa3_optics\data\reticles\bwa3_g36_reticle_ca.paa";
    reticleTextureSize = 1;

    bodyTexture = "\bwa3_optics\data\reticles\bwa3_g36_body_ca.paa";
    bodyTextureNight = "\bwa3_optics\data\reticles\bwa3_g36_body_night_ca.paa";
    bodyTextureSize = 1.95;
};

class CfgWeapons {
    class Rifle_Base_F;
    class BWA3_G36A2_base: Rifle_Base_F {
        class WeaponSlotsInfo;
    };

    class BWA3_G36A2: BWA3_G36A2_base {
        author = "$STR_BWA3_Author";

        // This entry is important so the PIP variant below will be replaced by this version when opening the Arsenal.
        baseWeapon = "BWA3_G36A2";

        // This sets the carry handle type. It defines a group of optics and which replacements the weapon should be able to switch to automatically. These replacements contain the carry handle optic mode of the weapon. The carry handle types are defined in `CBA_CarryHandleTypes`. (see below)
        CBA_CarryHandleType = "BWA3_G36_CarryHandle";

        // `CBA_ScriptedOptic` is the class that contains all the textures and values for the scripted 2d optics framework. Inherited from a base class, so we can edit the weapon and the replacement optics all in one place.
        class CBA_ScriptedOptic: BWA3_G36CarryHandleScriptedOptic_base {};

        // This weapon info type has to be set to `CBA_ScriptedOptic` or `CBA_ScriptedOptic_zooming` for the framework to trigger.
        weaponInfoType = "CBA_ScriptedOptic";

        // This defines the darkening around the scope. `big` or `small` can be used for a smaller or bigger diameter of the magnified area inside the scope. 70, 80, 90 and 100 for different values of transparency, with 100 being completely black and 70 being 30% transparency. Use your preference.
        modelOptics = "\bwa3_optics\bwa3_optics_reticle90_big.p3d";

        class OpticsModes {
            // This is the carry handle optic mode. Inherited so weapon and replacement optic can be edited in one place.
            class Scope: BWA3_G36CarryHandleScope_base {};

            // The iron sight mode. Nothing special. Will be replaced once an optic is attached alongside the carry handle mode. But due to the replacement optic variant that will be switched in by framework, a copy of the carry handle mode above will be available.
            class Kolimator: Scope {
                opticsID = 2;
                useModelOptics = 0;
                opticsPPEffects[] = {""};
                opticsFlare = 0;
                opticsDisablePeripherialVision = 0;
                opticsZoomMin = 0.25;
                opticsZoomMax = 1.25;
                opticsZoomInit = 0.75;
                memoryPointCamera = "eye";
                visionMode[] = {};
            };
        };

        // Inherits the compatible optics config defined above. Nothing special otherwise.
        class WeaponSlotsInfo: WeaponSlotsInfo {
            class CowsSlot: BWA3_CowsSlot_CarryHandle {
                iconPosition[] = {0.6,0.23};
            };
        };
    };

    // This is the PIP variant. Set `scope` to 1 so it doesn't appear in the arsenal or 
    class BWA3_G36A1_pip: BWA3_G36A1 {
        author = "$STR_BWA3_Author";
        scope = 1;

        // The optics model has to be changed to this one, or the PIP field around the scope will not be displayed. Can use "small" instead of "big" for optics with smaller diameter.
        class ItemInfo: ItemInfo {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
        };
    };

    // @ todo, explain optic
    class BWA3_optic_Kolimator_base: ItemCore {
        weaponInfoType = "RscWeaponZeroing";

        class ItemInfo: InventoryOpticsItem_Base_F {
            mass = 4;
            modelOptics = "\A3\Weapons_F\empty";
            optics = 1;

            class OpticsModes {
                class Kolimator {
                    opticsID = 1;
                    useModelOptics = 0;
                    opticsZoomMin = 0.25;
                    opticsZoomMax = 1.25;
                    opticsZoomInit = 0.75;
                    memoryPointCamera = "eye";
                    opticsFlare = 0;
                    opticsDisablePeripherialVision = 0;
                    distanceZoomMin = 200;
                    distanceZoomMax = 200;
                    cameraDir = "";
                    visionMode[] = {};
                    opticsPPEffects[] = {""};
                };
            };
        };

        inertia = 0;
    };

    class BWA3_optic_RSAS: BWA3_optic_Kolimator_base {
        author = "$STR_BWA3_Author";
        scope = 2;
    };

    class BWA3_optic_RSAS_G36A2: BWA3_optic_RSAS {
        author = "$STR_BWA3_Author";
        scope = 1;

        class CBA_ScriptedOptic: BWA3_G36CarryHandleScriptedOptic_base {};
        weaponInfoType = "RscWeaponZeroing";

        class ItemInfo: ItemInfo {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_90.p3d";

            class OpticsModes: OpticsModes {
                class Scope: BWA3_G36CarryHandleScope_base {};
                class Kolimator: Kolimator {};
            };
        };
    };

    class BWA3_optic_RSAS_G36A2_pip: BWA3_optic_RSAS_G36A2 {
        author = "$STR_BWA3_Author";
        scope = 1;

        class ItemInfo: ItemInfo {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
        };
    };
};
```

</details>

<details><summary>PIPItems and CarryHandleTypes configs</summary>

```cpp
class CBA_PIPItems {
    BWA3_G36A1 = "BWA3_G36A1_pip";
    BWA3_optic_RSAS_G36A2 = "BWA3_optic_RSAS_G36A2_pip";
};

class CBA_CarryHandleTypes {
    class BWA3_G36_CarryHandle {
        BWA3_optic_RSAS = "BWA3_optic_RSAS_G36A2";
    };
};
```

</details>

_Note that this is not the actual config of the BWA3 mod, but simplified for the purpose of this doc._

**Comment by commy2** [comment]
## Example configs

### Add PIP support to optic with scripted 2d reticle

<details><summary></summary>

```cpp
class CBA_PIPItems {
    BWA3_optic_ZO4x30 = "BWA3_optic_ZO4x30_pip";
};

class CfgWeapons {
    class ItemCore;
    class InventoryOpticsItem_Base_F;

    class BWA3_optic_ZO4x30: ItemCore {
        author = "$STR_BWA3_Author";
        scope = 2;

        class CBA_ScriptedOptic {
            reticleTexture = "\bwa3_optics\data\reticles\bwa3_zo4x30_reticle_ca.paa";
            reticleTextureSize = 1;

            bodyTexture = "\bwa3_optics\data\reticles\bwa3_zo4x30_body_black_ca.paa";
            bodyTextureNight = "\bwa3_optics\data\reticles\bwa3_zo4x30_body_night_ca.paa";
            bodyTextureSize = 1.95;
        };
        weaponInfoType = "CBA_ScriptedOptic";

        class ItemInfo: InventoryOpticsItem_Base_F {
            modelOptics = "\bwa3_optics\bwa3_optics_reticle90_big.p3d";
        };
    };

    class BWA3_optic_ZO4x30_pip: BWA3_optic_ZO4x30 {
        author = "$STR_BWA3_Author";
        scope = 1;

        class ItemInfo: ItemInfo {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
        };
    };
};
```

</details>

### Add PIP support to weapon with scripted 2d reticle

<details><summary></summary>

```cpp
class CBA_PIPItems {
    BWA3_G36A1 = "BWA3_G36A1_pip";
};

class CfgWeapons {
    class BWA3_G36A1_base;
    class BWA3_G36A1: BWA3_G36A1_base {
        author = "$STR_BWA3_Author";
        scope = 2;
        baseWeapon = "BWA3_G36A1";

        class CBA_ScriptedOptic {
            reticleTexture = "\bwa3_optics\data\reticles\bwa3_g36_reticle_ca.paa";
            reticleTextureSize = 1;

            bodyTexture = "\bwa3_optics\data\reticles\bwa3_g36_body_ca.paa";
            bodyTextureNight = "\bwa3_optics\data\reticles\bwa3_g36_body_night_ca.paa";
            bodyTextureSize = 1.95;
        };
        weaponInfoType = "CBA_ScriptedOptic";

        modelOptics = "\bwa3_optics\bwa3_optics_reticle90_big.p3d";
    };

    class BWA3_G36A1_pip: BWA3_G36A1 {
        author = "$STR_BWA3_Author";
        scope = 1;

        modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
    };
};
```

</details>

**Comment by commy2** [comment]
`"BWA3_Sniper_G82_Fleck"`
```sqf
0 spawn {
sleep 1;
isNil {
cameraOn addPrimaryWeaponItem "cba_2doptics_magnificationHelper_5";
{cameraOn addPrimaryWeaponItem "BWA3_optic_PMII_ShortdotCC_pip";} call CBA_fnc_execNextFrame;
}};
```

File: addons/2doptics/RscInGameUI.hpp [hunk 3]
```diff
@@ -13,6 +13,7 @@ class RscInGameUI {
         onLoad = QUOTE(call FUNC(loadScriptedOptic));
         controls[] = {"CA_FOVMode","ScopeBlack","Reticle","BodyNight","BodyDay","TrippleHeadLeft","TrippleHeadRight","CA_Zeroing","Magnification"};
 
+        // Idea by Taosenai. This control can be used to determine whether the scope or the kolimator is used.
```
Comment (dedmen) [inline addons/2doptics/RscInGameUI.hpp]: Collimator

**Comment by commy2** [comment]
## Example configs

### Add PIP support to weapon (with scripted 2d reticle)

<details><summary></summary>

```cpp
class CBA_PIPItems {
    BWA3_G36A1 = "BWA3_G36A1_pip";
};

class CfgWeapons {
    class BWA3_G36A1_base;
    class BWA3_G36A1: BWA3_G36A1_base {
        author = "$STR_BWA3_Author";
        scope = 2;
        baseWeapon = "BWA3_G36A1";

        // optional class, scripted 2d reticle data
        class CBA_ScriptedOptic {
            reticleTexture = "\bwa3_optics\data\reticles\bwa3_g36_reticle_ca.paa";
            reticleTextureSize = 1;

            bodyTexture = "\bwa3_optics\data\reticles\bwa3_g36_body_ca.paa";
            bodyTextureNight = "\bwa3_optics\data\reticles\bwa3_g36_body_night_ca.paa";
            bodyTextureSize = 1.95;
        };
        weaponInfoType = "CBA_ScriptedOptic";

        modelOptics = "\x\cba\addons\2doptics\cba_optic_big_90.p3d";
    };

    class BWA3_G36A1_pip: BWA3_G36A1 {
        author = "$STR_BWA3_Author";
        scope = 1;

        modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
    };
};
```

Notes:
- `baseWeapon` has to be set to the normal weapon class for this to properly work with the Arsenal.
- `scope` has to be set to private (= 1) for the pip variant of the weapon.
- `class CBA_ScriptedOptic` is optional and only needed if this should be a scripted 2d optic as well.
- The `weaponInfoType` has to be set to `CBA_ScriptedOptic` (or `CBA_ScriptedOptic_zooming`).
- The `modelOptics` has to have a face assigned with a procedural texture named `#(argb,512,512,1)r2t(cba_2doptics_rendertarget0,1)`, e.g. `\x\cba\addons\2doptics\cba_optic_small_pip.p3d` or `\x\cba\addons\2doptics\cba_optic_big_pip.p3d`, or a custom made model.

</details>

### Add PIP support to optic (with scripted 2d reticle)

<details><summary></summary>

```cpp
class CBA_PIPItems {
    BWA3_optic_ZO4x30 = "BWA3_optic_ZO4x30_pip";
};

class asdg_OpticRail;
class asdg_OpticRail1913: asdg_OpticRail {
    class compatibleItems {
        BWA3_optic_ZO4x30 = 1;
        BWA3_optic_ZO4x30_pip = 1;
    };
};

class CfgWeapons {
    // the weapon
    class Rifle_Base_F;
    class BWA3_G36A3_base: Rifle_Base_F {
        class WeaponSlotsInfo;
    };

    class BWA3_G36A3: BWA3_G36A3_base {
        class WeaponSlotsInfo: WeaponSlotsInfo {
            class MuzzleSlot: asdg_OpticRail1913 {};
        };
    };

    // the optic
    class ItemCore;
    class InventoryOpticsItem_Base_F;

    class BWA3_optic_ZO4x30: ItemCore {
        author = "$STR_BWA3_Author";
        scope = 2;

        // optional class, scripted 2d reticle data
        class CBA_ScriptedOptic {
            reticleTexture = "\bwa3_optics\data\reticles\bwa3_zo4x30_reticle_ca.paa";
            reticleTextureSize = 1;

            bodyTexture = "\bwa3_optics\data\reticles\bwa3_zo4x30_body_black_ca.paa";
            bodyTextureNight = "\bwa3_optics\data\reticles\bwa3_zo4x30_body_night_ca.paa";
            bodyTextureSize = 1.95;
        };
        weaponInfoType = "CBA_ScriptedOptic";

        class ItemInfo: InventoryOpticsItem_Base_F {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_90.p3d";
        };
    };

    class BWA3_optic_ZO4x30_pip: BWA3_optic_ZO4x30 {
        author = "$STR_BWA3_Author";
        scope = 1;

        class ItemInfo: ItemInfo {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
        };
    };
};
```

Notes:
- Both classes normal optic and pip variant have to be set as compatible optic for all weapons compatible with either for this to work properly.
- `scope` has to be set to private (= 1) for the pip variant of the optic.
- `class CBA_ScriptedOptic` is optional and only needed if this should be a scripted 2d optic as well.
- The `weaponInfoType` has to be set to `CBA_ScriptedOptic` (or `CBA_ScriptedOptic_zooming`).
- The `modelOptics` has to have a face assigned with a procedural texture named `#(argb,512,512,1)r2t(cba_2doptics_rendertarget0,1)`, e.g. `\x\cba\addons\2doptics\cba_optic_small_pip.p3d` or `\x\cba\addons\2doptics\cba_optic_big_pip.p3d`, or a custom made model.

</details>

### Add limited optic attachment support to weapons with carry handle optics (including pip support)

<details><summary></summary>

```cpp
class CBA_CarryHandleTypes {
    class BWA3_G36_CarryHandle {
        BWA3_optic_RSAS = "BWA3_optic_RSAS_G36A2";
    };
};

class CBA_PIPItems {
    BWA3_G36A2 = "BWA3_G36A2_pip";
    BWA3_optic_RSAS_G36A2 = "BWA3_optic_RSAS_G36A2_pip";
};

// compatible items info of the weapons optic slot
class BWA3_CowsSlot_CarryHandle {
    iconPicture = "\A3\Weapons_F\Data\UI\attachment_top.paa";
    iconPinpoint = "Bottom";
    iconPosition[] = {0.6,0.33};
    iconScale = 0.15;
    linkProxy = "\A3\data_f\proxies\weapon_slots\TOP";
    displayName = "$STR_A3_CowsSlot0";

    class compatibleItems {
        BWA3_optic_RSAS = 1;
        BWA3_optic_RSAS_G36A2 = 1;
        BWA3_optic_RSAS_G36A2_pip = 1;
    };
};

// shared optic mode of the carry handle optic
class BWA3_G36CarryHandleScope_base {
    opticsID = 1;
    useModelOptics = 1;
    opticsPPEffects[] = {"OpticsCHAbera5","OpticsBlur5"};
    opticsFlare = 1;
    opticsDisablePeripherialVision = 1;
    opticsZoomMin = "0.25/3";
    opticsZoomMax = "0.25/3";
    opticsZoomInit = "0.25/3";
    memoryPointCamera = "opticView";
    distanceZoomMin = 200;
    distanceZoomMax = 200;
    cameraDir = "";
    visionMode[] = {"Normal"};
};

// shared scripted 2d reticle data
class BWA3_G36CarryHandleScriptedOptic_base {
    reticleTexture = "\bwa3_optics\data\reticles\bwa3_g36_reticle_ca.paa";
    reticleTextureSize = 1;

    bodyTexture = "\bwa3_optics\data\reticles\bwa3_g36_body_ca.paa";
    bodyTextureNight = "\bwa3_optics\data\reticles\bwa3_g36_body_night_ca.paa";
    bodyTextureSize = 1.95;
};

class CfgWeapons {
    // the weapon with carry handle optic
    class Rifle_Base_F;
    class BWA3_G36A2_base: Rifle_Base_F {
        class WeaponSlotsInfo;
    };

    class BWA3_G36A2: BWA3_G36A2_base {
        CBA_CarryHandleType = "BWA3_G36_CarryHandle";

        author = "$STR_BWA3_Author";
        scope = 2;
        baseWeapon = "BWA3_G36A2";

        class CBA_ScriptedOptic: BWA3_G36CarryHandleScriptedOptic_base {};
        weaponInfoType = "CBA_ScriptedOptic";

        modelOptics = "\x\cba\addons\2doptics\cba_optic_big_90.p3d";

        class OpticsModes {
            class Scope: BWA3_G36CarryHandleScope_base {};
        };

        class WeaponSlotsInfo: WeaponSlotsInfo {
            class CowsSlot: BWA3_CowsSlot_CarryHandle {};
        };
    };

    // pip variant of the weapon with carry handle optic
    class BWA3_G36A2_pip: BWA3_G36A2 {
        author = "$STR_BWA3_Author";
        scope = 1;
        modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
    };

    // the cqb optic
    class ItemCore;
    class InventoryOpticsItem_Base_F;

    class BWA3_optic_RSAS: ItemCore {
        author = "$STR_BWA3_Author";
        scope = 2;

        class ItemInfo: InventoryOpticsItem_Base_F {
            class OpticsModes {
                class Kolimator {
                    // insert cqb optic mode config
                };
            };
        };
    };

    // the cqb optic with integrated carry handle optic mode
    class BWA3_optic_RSAS_G36A2: BWA3_optic_RSAS {
        author = "$STR_BWA3_Author";
        scope = 1;

        class CBA_ScriptedOptic: BWA3_G36CarryHandleScriptedOptic_base {};
        weaponInfoType = "RscWeaponZeroing";

        class ItemInfo: ItemInfo {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_90.p3d";

            class OpticsModes: OpticsModes {
                class Scope: BWA3_G36CarryHandleScope_base {};
                class Kolimator: Kolimator {};
            };
        };
    };

    // the cqb variant with the pip version of the integrated carry handle optic mode
    class BWA3_optic_RSAS_G36A2_pip: BWA3_optic_RSAS_G36A2 {
        author = "$STR_BWA3_Author";
        scope = 1;

        class ItemInfo: ItemInfo {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_pip.p3d";
        };
    };
};
```

</details>

### Add optic with zooming reticle

<details><summary></summary>

```cpp
class CfgWeapons {
    class ItemCore;
    class InventoryOpticsItem_Base_F;

    class BWA3_optic_PMII_ShortdotCC: ItemCore {
        class CBA_ScriptedOptic {
            minMagnificationReticleScale[] = {1.5,0.67};
            maxMagnificationReticleScale[] = {8,3.487};

            reticleDetailTextures[] = {
                {0, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_ca.paa", 1, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_illum_ca.paa"},
                {4, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_detail_ca.paa", .5, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_illum_detail_ca.paa"}
            };

            bodyTexture = "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-body_ca.paa";
            bodyTextureSize = 1.95;

            hideRedDotMagnification = 3.99;
            fadeReticleInterval[] = {3.5,2.5};
        };
        weaponInfoType = "CBA_ScriptedOptic_zooming";

        class ItemInfo: InventoryOpticsItem_Base_F {
            modelOptics = "\x\cba\addons\2doptics\cba_optic_big_90.p3d";

            class OpticsModes {
                class Scope {
                    // insert scope optic mode config
                    opticsZoomMin = "8 call (uiNamespace getVariable 'cba_2doptics_fnc_setOpticMagnificationHelper')";
                    opticsZoomMax = "1.1 call (uiNamespace getVariable 'cba_2doptics_fnc_setOpticMagnificationHelper')";
                    opticsZoomInit = "1.1 call (uiNamespace getVariable 'cba_2doptics_fnc_setOpticMagnificationHelper')";
                    discreteDistance[] = {100,200,300,400,500,600,700,800,900,1000,1100,1200};
                    discreteDistanceInitIndex = "2 call (uiNamespace getVariable 'cba_2doptics_fnc_setOpticMagnificationHelperZeroing')";
                };
            };
        };
    };
};
```

Notes:
- `cba_2doptics_fnc_setOpticMagnificationHelper` and `cba_2doptics_fnc_setOpticMagnificationHelperZeroing` are mandatory to keep magnification when switching between optic modes.
- `weaponInfoType` has to be set to `CBA_ScriptedOptic_zooming` instead of `CBA_ScriptedOptic`.

</details>

**Comment by commy2** [comment]
## `class CBA_ScriptedOptic` config entries
- The class is ignored unless `weaponInfoType` is set to `CBA_ScriptedOptic` or `CBA_ScriptedOptic_zooming`.

### Config entries for both `CBA_ScriptedOptic` and `CBA_ScriptedOptic_zooming`
- `bodyTexture`: texture of the scope body
- `bodyTextureNight`: texture of the scope body in darkness, optional (falls back to `bodyTexture`)
- `bodyTextureSize`: scope body texture scale factor (> 1 bigger, < 1 smaller)

### Config entries for `CBA_ScriptedOptic`
- `reticleTexture`: texture of the reticle
- `reticleTextureNight`: texture of the reticle in darkness, optional. Used for illuminated reticles.
- `reticleTextureSize`: reticle texture scale factor

### Config entries for `CBA_ScriptedOptic_zooming`
- `minMagnificationReticleScale[] = {minMagnification,minReticeScale};` and `maxMagnificationReticleScale[] = {maxMagnification,maxReticeScale};`
- These control how fast the reticle is scaled up when increasing the magnification, example:
```cpp
minMagnificationReticleScale[] = {5,1.026};
maxMagnificationReticleScale[] = {25,5*1.026};
```
- These mean the reticle texture is scaled to `1.026` at 5x magnification and to `5*1.026` at 25x magnification. Should be proportional. Not necessarily the minimum and maximum magnification of the optic, just two fixed points preferably far apart.

- `reticleDetailTextures[]`: texture LODs for different optic magnifications (to keep image sharp)
```cpp
reticleDetailTextures[] = {
    // {show at magnification > this value, reticleTexture, reticleTextureSize, reticleTextureNight (optional)}
    {0, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_ca.paa", 1, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_illum_ca.paa"},
    {4, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_detail_ca.paa", .5, "\bwa3_optics\data\reticles\bwa3_pmii_shortdotcc-reticle_illum_detail_ca.paa"}
};
```
- `hideRedDotMagnification`: Hide generic red dot texture above this magnification.
- `fadeReticleInterval[] = {start,end};`: Fade reticle from start magnification to completely disappear at end magnification.

**Comment by commy2** [comment]
## Scripted 2D optics

This framework provides support for scripted 2d optics, optics that have animated reticles when shooting the weapon or when changing the magnification of the optic continuously (as opposed to the discrete modes from the base game). 

This also includes "Picture in Picture" (PIP) optics. These are optics where the background around the scope is drawn unmagnified using the Arma 3 PIP technology. PIP optics are copies of normal 2d or 3d optics with a special optics model with a procedural "render to target" texture.

PIP variants of optics and weapons with integrated optics are hidden items that get swapped out automatically when the `Use picture in picture optics` setting is enabled locally. These classes are changed back when the setting is not enabled, thus you can change this setting during the mission and different players can enable or disable the setting without interfering with the preferences of others. The classes are also changed back when operating the Arsenal and are therefore compatible with the Save and Load functionality.

The for the PIP technology necessary camera object is not created when the `Use picture in picture optics` setting is disabled. Therefore you will not suffer any potential performance impact by this addition to CBA unless you make use of a compatible optic. Likewise, the rest of the framework is disabled and all script functions remain uncompiled when starting the game should none of your addons make use of this framework. CBA does not provide any optics and does not modify any optics from the base game either.

Another addition is the limited support for CQB optic attachments for weapons with integrated carry handle optics (e.g. H&K G36A2). By creating copies of CQB optics that include the weapon's carry handle optics mode, and swapping these optics in and out whenever necessary, it is possible to make these optics appear to work as expected and seamlessly to the players while remaining compatible with all other weapons.

File: addons/2doptics/fnc_initDisplayInterrupt.sqf [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+// Automatically hide when pause menu is closed.
+private _script = _display ctrlCreate ["RscMapControl", -1];
+
+_script ctrlSetPosition [0,0,0,0];
+_script ctrlCommit 0;
+
+_script ctrlAddEventHandler ["Draw", {
+    params ["_script"];
+    private _counter = ctrlParent _script getVariable "commy_fpsCounter";
```
Comment (Dystopian) [inline addons/2doptics/fnc_initDisplayInterrupt.sqf]: looks like forgotten debug

Comment (commy2) [reply addons/2doptics/fnc_initDisplayInterrupt.sqf]: Not debug, but doesn't belong.

Comment (commy2) [reply addons/2doptics/fnc_initDisplayInterrupt.sqf]: https://github.com/CBATeam/CBA_A3/pull/1118

**Pull request merged**